cpp/src/arrow/compute/util.cc

```diff
@@ -299,7 +313,22 @@ void bytes_to_bits(int64_t hardware_flags, const int num_bits, const uint8_t* by
   }
   int tail = num_bits % unroll;
   if (tail) {
-    uint64_t bytes_next = SafeLoadUpTo8Bytes(bytes + num_bits - tail, tail);
+    uint64_t bytes_next;
+#if ARROW_LITTLE_ENDIAN
+    bytes_next = SafeLoadUpTo8Bytes(bytes + num_bits - tail, tail);
+#else
+    if (tail == 8) {
```

Comment:
Does this happen?
I think that `tail` must not be `8` because `unroll` is always `8` and `tail = num_bits % unroll`.